src/Init/Core.lean

```diff
@@ -167,13 +167,24 @@ protected def bind {α : Type u} {β : Type v} (x : Task α) (f : α → Task β
 
 end Task
 
-/-- Some type that is not a scalar value in our runtime. -/
-structure NonScalar where
-  val : Nat
+private inductive PNonScalarSpec : Type u where
+  | mk (v : Nat)
+
+opaque nonScalarSpec : NonemptyType.{u} := ⟨PNonScalarSpec, ⟨⟨0⟩⟩⟩
```

Comment:
```suggestion
opaque nonScalarSpec : NonemptyType.{u}
```

Reply:
That doesn't look like a behavior-preserving change, since it will implement `NonScalar` as `Unit` instead of `Nat`. Is `Unit` actually a non-scalar type?

Reply:
At least currently, this doesn't make any difference since the compiler doesn't unfold opaque types.  We use this pattern quite extensively (just grep for `opaque.*Nonempty`).

With the new compiler, we might want to change that and better take advantage of the types.  But then `⟨PNonScalarSpec, ⟨⟨0⟩⟩⟩` would be just as wrong (because `PNonScalarSpec` can only contain scalar and mpz values, but `PNonScalar` should obviously contain more).

Come to think of it, the name `PNonScalar` is extremely misleading.  We use it not just for non-scalar values, but for *potentially* non-scalar values (i.e. including scalars).

Reply:
I believe *scalar* here means "unboxed integer type", i.e. not `Nat`

Reply:
> At least currently, this doesn't make any difference since the compiler doesn't unfold opaque types. We use this pattern quite extensively (just grep for `opaque.*Nonempty`).

I'm aware that we use `opaque.*Nonempty` a lot, but my understanding was that this was for blocking the kernel from seeing through the type alias while still letting the compiler use the same runtime representation. That is, I would have expected an opaque wrapper around `UInt8` to still have a `uint8_t` ABI.

Reply:
> Come to think of it, the name `PNonScalar` is extremely misleading. We use it not just for non-scalar values, but for _potentially_ non-scalar values (i.e. including scalars).

Everything should be "boxed" in this type, including scalars. That may not mean that there is actually a pointer indirection because of the nat optimization, but even if the value is "actually" a `UInt8` it should still be passed as a `lean_object*`. So I don't think the name is that misleading, although it is subtle and perhaps another name is more suitable (what's a scalar anyway? we don't use this terminology anywhere else and we use it incompatibly in mathlib).

Reply:
I think a better name might be something like `OpaqueBox` or `OpaquePointer`. Something more evocative of `void*`.

Reply:
I got the scalar name from `lean_is_scalar`.  It is indeed confusing if the terminology is used for unboxed values as well.

I would suggest to rename `NonScalar` to `Object` for consistency with `ObjectMap`.

Reply:
We also have `tobject/object` in https://leanprover-community.github.io/mathlib4_docs/Lean/Compiler/IR/Basic.html#Lean.IR.IRType

Reply:
I'm not a fan of grabbing such prime real estate as `_root_.Object` for this niche thing. Can we at least namespace it? Or use `LCObject` to go with the other `lcFunction`s.

Reply:
Fwiw it's `Std.ObjectMap` so `Std.Object` would be consistent.

Reply:
`Std.ObjectMap` is *in* Std though. `NonScalar` is in `Init` because it is used by some array ops.

Reply:
`Std.Format` is also in init. `Lean.Syntax` too.